src/main/java/ch/njol/skript/expressions/ExprHash.java

```diff
@@ -1,112 +1,76 @@
+		"These hashing algorithms are not suitable for hashing passwords.",
+		"If handling passwords, use a <a href='https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#password-hashing-algorithms'>hashing algorithm specifically designed for passwords</a>.",
+		"MD5 is deprecated and may be removed in a future release. It is provided mostly for backwards compatibility, as it is outdated and not secure. ",
+		"SHA is more secure, but is not suitable for hashing passwords (even with salting). ",
 		"When hashing data, you <strong>must</strong> specify algorithms that will be used for security reasons! ",
-		"<p>Please note that a hash cannot be reversed under normal circumstanses. You will not be able to get original value from a hash with Skript."})
-@Examples({
-		"command /setpass &lt;text&gt;:",
-		"\ttrigger:",
-		"\t\tset {password::%uuid of player%} to text-argument hashed with SHA-256",
-		"command /login &lt;text&gt;:",
-		"\ttrigger:",
-		"\t\tif text-argument hashed with SHA-256 is {password::%uuid of player%}:",
-		"\t\t\tmessage \"Login successful.\"",
-		"\t\telse:",
-		"\t\t\tmessage \"Wrong password!\""})
-@Since("2.0, 2.2-dev32 (SHA-256 algorithm)")
+		"Please note that a hash cannot be reversed under normal circumstances. You will not be able to get original value from a hash with Skript."
+})
+@Example("set {_hash} to \"hello world\" hashed with SHA-256")
+@Since("2.0, 2.2-dev32 (SHA-256 algorithm), INSERT VERSION (SHA-384, SHA-512)")
 public class ExprHash extends PropertyExpression<String, String> {
+
+	private static final HexFormat HEX_FORMAT = HexFormat.of().withLowerCase();
+
 	static {
-		Skript.registerExpression(ExprHash.class, String.class, ExpressionType.SIMPLE,
-				"%strings% hash[ed] with (0¦MD5|1¦SHA-256)");
-	}
-	
-	@SuppressWarnings("null")
-	private final static Charset UTF_8 = Charset.forName("UTF-8");
-	
-	@Nullable
-	static MessageDigest md5;
-	@Nullable
-	static MessageDigest sha256;
-	
-	static {
-		try {
-			md5 = MessageDigest.getInstance("MD5");
-			sha256 = MessageDigest.getInstance("SHA-256");
-		} catch (final NoSuchAlgorithmException e) {
-			throw new InternalError("JVM does not adhere to Java specifications");
-		}
+		Skript.registerExpression(ExprHash.class, String.class, ExpressionType.COMBINED,
+				"%strings% hash[ed] with (:(MD5|SHA-256|SHA-384|SHA-512))");
 	}
+
+	private MessageDigest digest;
 	
-	private int algorithm;
-	
-	@SuppressWarnings({"unchecked", "null"})
 	@Override
-	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		//noinspection unchecked
 		setExpr((Expression<? extends String>) exprs[0]);
-		algorithm = parseResult.mark;
-		return true;
+		String algorithm = parseResult.tags.get(0).toUpperCase();
```

Comment:
Not sure if this is actually required (since the syntax is uppercase, meaning the parse tag should be too), but we'll want to specify the locale
```suggestion
		String algorithm = parseResult.tags.get(0).toUpperCase(Locale.ENGLISH);
```